asf_public_discourse_home_decarbonisation/analysis/buildhub/extract_faqs.py

```diff
@@ -0,0 +1,245 @@
+
+    This can be particularly useful for analysing text where it's important to both identify expressions of uncertainty and examine the context without these expressions.
+
+    Args:
+        sentences (List[str]): A list of sentences to be evaluated. Each item in the list should be a string representing a single sentence.
+
+    Returns:
+        Tuple[List[str], List[str]]:
+            - The first list contains sentences that include any of the predefined "I don't know" phrases.
+            - The second list contains the same sentences but with the "I don't know" phrases removed, allowing for analysis of the remaining sentence content.
+
+    Note:
+        - The function is case-insensitive when searching for the predefined phrases.
+        - The removal of phrases is done in a straightforward manner, which may not account for complex sentence structures or punctuation.
+    """
+    idk_phrases = []
+    sentences_without_inclusion = []
+    inclusion_phrases = [
+        "i don’t know ",
+        "i do not know ",
+        "i do not know how to ",
+    ]  # List of phrases to include
+    for sentence in sentences:
+        if any(phrase in sentence.lower() for phrase in inclusion_phrases):
+            idk_phrases.append(sentence)
+            for phrase in inclusion_phrases:
+                sentence = sentence.lower().replace(phrase, "")
+            sentences_without_inclusion.append(sentence)
+    return idk_phrases, sentences_without_inclusion
+
+
+if __name__ == "__main__":
+    # Read the CSV file
+    args = create_argparser()
+    category = args.category
+    collection_date_time = args.collection_date_time
+    # uses different getter function depending on the forum
+    if args.forum == "bh":
+        category_dataframe = get_bh_category_data(category, collection_date_time)
+    elif args.forum == "mse":
+        category_dataframe = get_mse_category_data(category, collection_date_time)
+    else:
+        logger.info("Please enter a valid forum name (bh or mse)")
+        sys.exit(-1)
+    FORUM_FAQ_PATH = os.path.join(
+        PROJECT_DIR, f"outputs/extracted_questions/{args.forum}/forum_{category}/"
+    )
+    # Ensure the output directory exists
+    os.makedirs(FORUM_FAQ_PATH, exist_ok=True)
+    # Tokenize sentences
+    nltk.download("punkt")
+    # Convert all text to strings to avoid type errors
+    category_dataframe["text"] = category_dataframe["text"].astype(str)
+    category_dataframe["sentences"] = category_dataframe["text"].apply(sent_tokenize)
+    # Apply the function to extract questions
+    category_dataframe["questions"] = category_dataframe["sentences"].apply(
+        extract_questions
+    )
+    (
+        category_dataframe["idk_phrases"],
+        category_dataframe["sentences_without_inclusion"],
+    ) = zip(*category_dataframe["sentences"].apply(extract_idk))
```

Comment:
This could also be in a function that then is called here, something like `identify_questions_and_dont_know_patterns()`

Reply:
Do we feel like this adds much? Think it might be nice to be explicit/clear here about what each line does instead of a function which calls a bunch of different functions? Might be wrong on this though! I've changed this now to be `    # Apply the function to extract questions
    category_dataframe["questions"] = category_dataframe["sentences"].apply(
    lambda sentences: extract_questions(sentences, question_words)
    )
    # Apply the function to filter out short questions
    category_dataframe["questions"] = category_dataframe["questions"].apply(
    lambda questions: filter_out_short_questions(questions)
    )
    # Apply the function to extract "I don't know" phrases
    category_dataframe["idk_phrases"] = category_dataframe["sentences"].apply(
    lambda sentences: extract_idk(sentences, inclusion_phrases)
    )
    # Apply the function to extract "I don't know" phrases
    category_dataframe["sentences_without_inclusion"] = category_dataframe["idk_phrases"].apply(
    lambda idk_phrases: extract_X_from_idk(idk_phrases, inclusion_phrases)
    )`

Reply:
sounds good!